spec/GraphQLOverHTTP.md

```diff
@@ -237,120 +311,304 @@ omitted.
 
-The following examples provide guidance on how to deal with specific error
-cases:
+Note: Using `4xx` and `5xx` status codes in this situation is not recommended -
+since no {GraphQL request error} has occurred it is seen as a "partial
+response".
 
-### Unparseable or invalid request body
+If the GraphQL response does not contain the {data} entry then the server MUST
+reply with a `4xx` or `5xx` status code as appropriate.
 
-For example: `NONSENSE`, `{"qeury": "{__typena`
+Note: The GraphQL specification indicates that the only situation in which the
+GraphQL response does not include the {data} entry is one in which the {errors}
+entry is populated.
 
-Completely prevents execution of the GraphQL operation and SHOULD result in
-status code `400` (Bad Request).
+If the GraphQL request is invalid (e.g. it is malformed, or does not pass
+validation) then the server SHOULD reply with `400` status code.
 
-### Document validation
+If the client is not permitted to issue the GraphQL request then the server
+SHOULD reply with `403`, `401` or similar appropriate status code.
 
-Includes validation steps that run before execution of the GraphQL operation:
+#### Examples
 
-- [GraphQL specification validation](https://spec.graphql.org/October2021/#sec-Validation)
-- custom validation, for example: depth limit, complexity limit
+The following examples provide guidance on how to deal with specific error cases
+when using the `application/graphql+json` media type to encode the response
+body:
 
-The server SHOULD deny execution with a status code of `400` (Bad Request).
+##### Invalid request body or parsing failure
 
-Note: In certain circumstances, for example persisted queries that were
+For example: `NONSENSE`, `{"qeury": "{__typena`
+
+Requests that do not constitute a well-formed GraphQL request should result in
+status code `400` (Bad Request).
+
+##### Document validation failure
+
+Requests that fail to pass validation SHOULD be denied execution with a status
+code of `400` (Bad Request).
+
+Note: In certain circumstances, for example persisted operations that were
 previously known to be valid, the server MAY attempt execution regardless of
 validation errors.
 
-### Runtime validation
+Note: Validation rules include those specified in
+[the Validation section of the GraphQL specification](http://spec.graphql.org/draft/#sec-Validation),
+and any custom validation rules the server is using (for example: depth limit,
+complexity limit).
+
+##### Operation cannot be determined
+
+If [GetOperation()](<https://spec.graphql.org/draft/#GetOperation()>) raises a
+{GraphQL request error}, the server SHOULD NOT execute the request and SHOULD
+return a status code of `400` (Bad Request).
+
+##### Variable coercion failure
+
+If
+[CoerceVariableValues()](<https://spec.graphql.org/draft/#CoerceVariableValues()>)
+raises a {GraphQL request error}, the server SHOULD NOT execute the request and
+SHOULD return a status code of `400` (Bad Request).
```

Comment:
One note: while this is a good idea which I support (and which Apollo Server tries to implement), it's challenging to implement this today in a non-hacky way with `graphql-js`. I thought there was already an issue for this but I couldn't find one so I filed https://github.com/graphql/graphql-js/issues/3679

Reply:
Yeah; I believe the trick right now is that if it doesn't have a `data` property (note that's different from it being present but null) then you can assume it's a `CoerceVariableValues()` or similar request error as opposed to a field error. I think you can actually just straight up use `graphql(...)` and use this heuristic to adhere to the GraphQL-over-HTTP spec, you don't need to call out to the separate methods since all the request errors end up in HTTP 400 anyway. But yeah, it'd be beneficial to have it separate for sure :+1:

Reply:
Ah true, I believe we do map lack of data to 400 but we also want a more
precise error code inline for the different error types and that's what we
can't do without better support from graphql-js.

Reply:
💯